java/memory/src/main/java/org/apache/arrow/memory/util/hash/MurmurHasher.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util.hash;
+
+import static io.netty.util.internal.PlatformDependent.getByte;
+import static io.netty.util.internal.PlatformDependent.getInt;
+
+import org.apache.arrow.memory.util.ArrowBufPointer;
+
+import io.netty.buffer.ArrowBuf;
+
+/**
+ * Implementation of the Murmur hashing algorithm.
+ * Details of the algorithm can be found in
+ * https://en.wikipedia.org/wiki/MurmurHash
+ * <p>
+ *   Murmur hashing is computationally expensive, as it involves several
+ *   integer multiplications. However, the produced hash codes have
+ *   good quality in the sense that they are uniformly distributed in the universe.
+ * </p>
+ * <p>
+ *   Therefore, this algorithm is suitable for scenarios where uniform hashing
+ *   is desired (e.g. in an open addressing hash table/hash set).
+ * </p>
+ */
+public class MurmurHasher implements ArrowBufHasher {
+
+  private final int seed;
+
+  /**
+   * Creates a default Murmur hasher, with seed 0.
+   */
+  public MurmurHasher() {
+    this(0);
+  }
+
+  /**
+   * Creates a Murmur hasher.
+   * @param seed the seed for the hasher.
+   */
+  public MurmurHasher(int seed) {
+    this.seed = seed;
+  }
+
+  @Override
+  public int hashCode(long address, int length) {
+    int index = 0;
+    int hash = seed;
+    while (index + 4 <= length) {
+      int intValue = getInt(address + index);
+      if (!ArrowBufPointer.LITTLE_ENDIAN) {
+        intValue = Integer.reverseBytes(intValue);
+      }
+      hash = combineHashCode(hash, intValue);
+      index += 4;
+    }
+
+    if (index < length) {
+      // process remaining data as a integer in little endian
+      int intValue = 0;
+      for (int i = index - 1; i >= index; i--) {
+        intValue <<= 8;
+        intValue |= (getByte(address + i) & 0x000000ff);
+        index += 1;
+      }
+      hash = combineHashCode(hash, intValue);
+    }
+    return finalizeHashCode(hash, length);
+  }
+
+  @Override
+  public int hashCode(ArrowBuf buf, int offset, int length) {
+    buf.checkBytes(offset, offset + length);
+    return hashCode(buf.memoryAddress() + offset, length);
+  }
+
+  /**
+   * Combine the current hash code and a new int value to calculate
+   * a new hash code.
+   * @param currentHashCode the current hash code.
+   * @param intValue the new int value.
+   * @return the new hah code.
+   */
+  public static int combineHashCode(int currentHashCode, int intValue) {
+    int c1 = 0xcc9e2d51;
+    int c2 = 0x1b873593;
+    int r1 = 15;
+    int r2 = 13;
+    int m = 5;
+    int n = 0xe6546b64;
+
+    int k = intValue;
+    k = k * c1;
+    k = k << r1;
```

Comment:
Per the sample C-code in the wiki page, this should be a rotation, not just a left shift.

k = (k << r1) | (k >> (32 - r1))

Reply:
@pravindra thanks for your careful review.
There is a consistency in the wiki page: the rotation above exists in the sample C code, but does not exist in the pseudo code.
I checked some other implementations, one from MIT: 
http://web.mit.edu/freebsd/head/sys/libkern/murmur3_32.c
and the other from Flink:
https://github.com/apache/flink/blob/master/flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/utils/MurmurHash.java

These implementations do not have the above rotation, so let's also omit it.
What do you think?

Reply:
Both of them do the rotate, I think.

in the MIT code, I see
  #define rol32(i32, n) ((i32) << (n) | (i32) >> (32 - (n)))

in the flink code, i see 
  Integer.rotateLeft(input, 15)

you can also see the impl in dremio-oss using arrow bufs 

https://github.com/dremio/dremio-oss/blob/master/sabot/kernel/src/main/java/com/dremio/exec/expr/fn/impl/MurmurHash3.java#L156

Reply:
I see. Thanks again for your careful review.